posts/ngrx-facade-pattern.md

```diff
@@ -0,0 +1,154 @@
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+After moving everything related to the store into our service, we can then inject that service into our component.
+
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private accountService: AccountService) {}
+
+  ngOnInit() {
+    this.accountList = this.accountService.accountList;
+
+    this.accountService.loadAccountList();
+  }
+
+  addAccount(account: string) {
+    this.accountService.addAccount(account);
+  }
+
+  editAccount(id: string, account: string) {
+    this.accountService.editAccount( id, account);
+  }
+
+  deleteAccount(id: string) {
+    this.accountService.deleteAccount(id);
+  }
+}
+
+```
+After implementing the facade pattern what did we actually gain?
+
+### Pros
+- Better developer experience through abstraction.
+    - You refactor your service or component without touching both.
+    - Your component doesn't care what library you use for managing state.
+- This service can be reused by other components that need to access our `accountList` instead of duplicating code.
+- Your components are more simplistic.
+    - Devs have a clearer focus on component functionality instead of NgRx stores, actions, and reducers.
+
+
+### Cons
+- Loss of indirection violates a redux pattern.
+    -  Indirection decouples "what happened" from the logic that determines "how the state should be updated". This allows redux to track how the state changed every time an action gets dispatched. 
+- At the cost of being easier to interact with, the facade pattern obfuscates how NgRx works. This may hinder the understanding of the state management lifecycle for other developers on the team.
+- Actions could be reused instead of treating them as unique events.
+   - If actions are reused, then changing one can impact multiple components across our application.
+   - The concept of reusing actions goes against [good action hygiene](https://indepth.dev/posts/1407/force-good-action-hygiene-and-write-less-actions-in-ngrx-with-the-prepared-events-pattern#:~:text=Good%20Action%20Hygiene,-Link%20to%20this&text=Takeaway%20%2D%20Treat%20Actions%20as%20unique,by%20going%20through%20the%20reducers.)
+   - [NgRx DevTools](https://blog.angular-university.io/angular-ngrx-devtools/) has an action log that may be less useful for debugging with generic actions.
+
+### Conclusion
+While this may violate some of the standards and recommendations of the redux pattern, I don't think the gains your team can have in development speed are outweighed. If you need to move fast there is a clear upside to using facades, but I would make sure the developers on the team also gain a solid understanding of NgRx. By understanding NgRx and the potential issues that using a facade may bring, you can write better code as a team and avoid those problems.
```

Comment:
```suggestion
While this may violate some of the standards and recommendations of the Redux pattern, I do not think the gains your team can have in development speed are outweighed. If you need to move fast, there is a clear upside to using facades, but I would make sure the developers on the team also gain a solid understanding of NgRx. By understanding NgRx and the potential issues that using a facade may bring, you can write better code as a team and avoid those problems.
```